splinterd/src/config/builder.rs

```diff
@@ -12,14 +12,441 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::config::PartialConfig;
+use std::collections::HashMap;
+use std::time::Duration;
+
+use crate::config::error::ConfigError;
+use crate::config::{Config, ConfigSource, PartialConfig};
 
 pub trait PartialConfigBuilder {
     /// Takes all values set in a config object to create a PartialConfig object.
     ///
     fn build(self) -> PartialConfig;
 }
 
+/// ConfigBuilder collects PartialConfig objects from various sources to be used to generate a
+/// Config object. The ConfigBuilder builds the Config object in a specific order determined by the
+/// ConfigSource of the PartialConfig.
+pub struct ConfigBuilder {
+    partial_configs: HashMap<ConfigSource, PartialConfig>,
+    storage: Option<(String, ConfigSource)>,
+    transport: Option<(String, ConfigSource)>,
+    cert_dir: Option<(String, ConfigSource)>,
+    ca_certs: Option<(String, ConfigSource)>,
+    client_cert: Option<(String, ConfigSource)>,
+    client_key: Option<(String, ConfigSource)>,
+    server_cert: Option<(String, ConfigSource)>,
+    server_key: Option<(String, ConfigSource)>,
+    service_endpoint: Option<(String, ConfigSource)>,
+    network_endpoint: Option<(String, ConfigSource)>,
+    peers: Option<(Vec<String>, ConfigSource)>,
+    node_id: Option<(String, ConfigSource)>,
+    bind: Option<(String, ConfigSource)>,
+    #[cfg(feature = "database")]
+    database: Option<(String, ConfigSource)>,
+    registry_backend: Option<(String, ConfigSource)>,
+    registry_file: Option<(String, ConfigSource)>,
+    heartbeat_interval: Option<(u64, ConfigSource)>,
+    admin_service_coordinator_timeout: Option<(Duration, ConfigSource)>,
+    state_dir: Option<(String, ConfigSource)>,
+}
+
+impl ConfigBuilder {
+    pub fn new() -> Self {
+        ConfigBuilder {
+            partial_configs: HashMap::new(),
+            storage: None,
```

Comment:
There should be no need for these fields, only the partial configs should be used when building the config result and could be localized to that method.

Reply:
Addressed.